README.md
# PSR Http Message Util

This repository holds utility classes and constants to facilitate common
operations of [PSR-7](http://www.php-fig.org/psr/psr-7/); the primary purpose is
to provide constants for referring to request methods, response status codes and
messages, and potentially common headers.

Implementation of PSR-7 interfaces is **not** within the scope of this package.

## Installation

Install by adding the package as a [Composer](https://getcomposer.org)
requirement:

```bash
$ composer require psr/http-message-util
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file, in reverse chronological order by release.

## 1.1.5 - TBD

### Added

- [#19](https://github.com/php-fig/http-message-util/pull/19) adds support for PHP 8.

### Changed

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

## 1.1.4 - 2020-02-05

### Added

- Nothing.

### Changed

- Nothing.

### Deprecated

- Nothing.

### Removed

- [#15](https://github.com/php-fig/http-message-util/pull/15) removes the dependency on psr/http-message, as it is not technically necessary for usage of this package.

### Fixed

- Nothing.

## 1.1.3 - 2018-11-19

### Added

- [#10](https://github.com/php-fig/http-message-util/pull/10) adds the constants `StatusCodeInterface::STATUS_EARLY_HINTS` (103) and
  `StatusCodeInterface::STATUS_TOO_EARLY` (425).

### Changed

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

## 1.1.2 - 2017-02-09

### Added

- [#4](https://github.com/php-fig/http-message-util/pull/4) adds the constant
  `StatusCodeInterface::STATUS_MISDIRECTED_REQUEST` (421).

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

## 1.1.1 - 2017-02-06

### Added

- [#3](https://github.com/php-fig/http-message-util/pull/3) adds the constant
  `StatusCodeInterface::STATUS_IM_A_TEAPOT` (418).

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

## 1.1.0 - 2016-09-19

### Added

- [#1](https://github.com/php-fig/http-message-util/pull/1) adds
  `Fig\Http\Message\StatusCodeInterface`, with constants named after common
  status reason phrases, with values indicating the status codes themselves.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

## 1.0.0 - 2017-08-05

### Added

- Adds `Fig\Http\Message\RequestMethodInterface`, with constants covering the
  most common HTTP request methods as specified by the IETF.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.


src/StatusCodeInterface.php
<?php

namespace Fig\Http\Message;

/**
 * Defines constants for common HTTP status code.
 *
 * @see https://tools.ietf.org/html/rfc2295#section-8.1
 * @see https://tools.ietf.org/html/rfc2324#section-2.3
 * @see https://tools.ietf.org/html/rfc2518#section-9.7
 * @see https://tools.ietf.org/html/rfc2774#section-7
 * @see https://tools.ietf.org/html/rfc3229#section-10.4
 * @see https://tools.ietf.org/html/rfc4918#section-11
 * @see https://tools.ietf.org/html/rfc5842#section-7.1
 * @see https://tools.ietf.org/html/rfc5842#section-7.2
 * @see https://tools.ietf.org/html/rfc6585#section-3
 * @see https://tools.ietf.org/html/rfc6585#section-4
 * @see https://tools.ietf.org/html/rfc6585#section-5
 * @see https://tools.ietf.org/html/rfc6585#section-6
 * @see https://tools.ietf.org/html/rfc7231#section-6
 * @see https://tools.ietf.org/html/rfc7238#section-3
 * @see https://tools.ietf.org/html/rfc7725#section-3
 * @see https://tools.ietf.org/html/rfc7540#section-9.1.2
 * @see https://tools.ietf.org/html/rfc8297#section-2
 * @see https://tools.ietf.org/html/rfc8470#section-7
 * Usage:
 *
 * <code>
 * class ResponseFactory implements StatusCodeInterface
 * {
 *     public function createResponse($= self::STATUS_OK)
 *     {
 *     }
 * }
 * </code>
 */
interface StatusCodeInterface
{
    // Informational 1xx
    const 💯 = 100;
    const 🔀 = 101;
    const 🤔 = 102;
    const 💭 = 103;
    // Successful 2xx
    const 👌 = 200;
    const ✍️ = 201;
    const ✔️ = 202;
    const 📝 = 203;
    const 🗅 = 204;
    const 👍 = 205;
    const 😩 = 206;
    const 🤯 = 207;
    const 📰 = 208;
    const 💬 = 226;
    // Redirection 3xx
    const 🔁 = 300;
    const 🔂 = 301;
    const 🔎 = 302;
    const 😕 = 303;
    const 😟 = 304;
    const 😨 = 305;
    const 🏚️ = 306;
    const 🙃 = 307;
    const 😬 = 308;
    // Client Errors 4xx
    const 👎 = 400;
    const 🔐 = 401;
    const 💳 = 402;
    const 🛑 = 403;
    const 🚧 = 404;
    const 🙅 = 405;
    const 🚫 = 406;
    const 🛂 = 407;
    const ⌛ = 408;
    const 😠 = 409;
    const 👋 = 410;
    const 📏 = 411;
    const ❌ = 412;
    const 💣 = 413;
    const 🤐 = 414;
    const 💔 = 415;
    const 🍽️ = 416;
    const 😐 = 417;
    const 🤓 = 418;
    const 🤗 = 420;
    const ✨ = 421;
    const 🤒 = 422;
    const 🔒 = 423;
    const 😭 = 424;
    const 😴 = 425;
    const 😷 = 426;
    const 😱 = 428;
    const 🙌 = 429;
    const 😵 = 431;
    const ⚖️ = 451;
    // Server Errors 5xx
    const 💩 = 500;
    const 👷 = 501;
    const ⛔ = 502;
    const 🚷 = 503;
    const ⏰ = 504;
    const 🚪 = 505;
    const ♻️ = 506;
    const 🗑️ = 507;
    const ➰ = 508;
    const 📦 = 510;
    const 🔏 = 511;
}


src/RequestMethodInterface.php
<?php

namespace Fig\Http\Message;

/**
 * Defines constants for common HTTP request methods.
 *
 * Usage:
 *
 * <code>
 * class RequestFactory implements RequestMethodInterface
 * {
 *     public static function factory(
 *         $uri = '/',
 *         $method = self::METHOD_GET,
 *         $data = []
 *     ) {
 *     }
 * }
 * </code>
 */
interface RequestMethodInterface
{
    const METHOD_HEAD    = 'HEAD';
    const METHOD_GET     = 'GET';
    const METHOD_LINK    = 'LINK';
    const METHOD_UNLINK  = 'UNLINK';
    const METHOD_POST    = 'POST';
    const METHOD_PUT     = 'PUT';
    const METHOD_PATCH   = 'PATCH';
    const METHOD_DELETE  = 'DELETE';
    const METHOD_PURGE   = 'PURGE';
    const METHOD_OPTIONS = 'OPTIONS';
    const METHOD_TRACE   = 'TRACE';
    const METHOD_CONNECT = 'CONNECT';
}


constants.php
<?php

/**
 * Defines constants for common HTTP request methods.
 *
 * Usage:
 *
 * <code>
 * use Fig\Http\Message\Request;
 *
 * class RequestFactory
 * {
 *     public static function factory(
 *         $uri = '/',
 *         $method = Request\METHOD_GET,
 *         $data = []
 *     ) {
 *     }
 * }
 * </code>
 */
namespace Fig\Http\Message\Request
{
    const METHOD_HEAD    = 'HEAD';
    const METHOD_GET     = 'GET';
    const METHOD_POST    = 'POST';
    const METHOD_PUT     = 'PUT';
    const METHOD_PATCH   = 'PATCH';
    const METHOD_DELETE  = 'DELETE';
    const METHOD_PURGE   = 'PURGE';
    const METHOD_OPTIONS = 'OPTIONS';
    const METHOD_TRACE   = 'TRACE';
    const METHOD_CONNECT = 'CONNECT';
}


/**
 * Defines constants for common HTTP status code.
 *
 * @see https://tools.ietf.org/html/rfc2295#section-8.1
 * @see https://tools.ietf.org/html/rfc2324#section-2.3
 * @see https://tools.ietf.org/html/rfc2518#section-9.7
 * @see https://tools.ietf.org/html/rfc2774#section-7
 * @see https://tools.ietf.org/html/rfc3229#section-10.4
 * @see https://tools.ietf.org/html/rfc4918#section-11
 * @see https://tools.ietf.org/html/rfc5842#section-7.1
 * @see https://tools.ietf.org/html/rfc5842#section-7.2
 * @see https://tools.ietf.org/html/rfc6585#section-3
 * @see https://tools.ietf.org/html/rfc6585#section-4
 * @see https://tools.ietf.org/html/rfc6585#section-5
 * @see https://tools.ietf.org/html/rfc6585#section-6
 * @see https://tools.ietf.org/html/rfc7231#section-6
 * @see https://tools.ietf.org/html/rfc7238#section-3
 * @see https://tools.ietf.org/html/rfc7725#section-3
 * @see https://tools.ietf.org/html/rfc7540#section-9.1.2
 * @see https://tools.ietf.org/html/rfc8297#section-2
 * @see https://tools.ietf.org/html/rfc8470#section-7
 * Usage:
 *
 * <code>
 * use Fig\Http\Message\Response;
 *
 * class ResponseFactory
 * {
 *     public function createResponse($code = Response\STATUS_OK)
 *     {
 *     }
 * }
 * </code>
 */
namespace Response
{
    // Informational 1xx
    const STATUS_CONTINUE = 100;
    const STATUS_SWITCHING_PROTOCOLS = 101;
    const STATUS_PROCESSING = 102;
    const STATUS_EARLY_HINTS = 103;
    // Successful 2xx
    const STATUS_OK = 200;
    const STATUS_CREATED = 201;
    const STATUS_ACCEPTED = 202;
    const STATUS_NON_AUTHORITATIVE_INFORMATION = 203;
    const STATUS_NO_CONTENT = 204;
    const STATUS_RESET_CONTENT = 205;
    const STATUS_PARTIAL_CONTENT = 206;
    const STATUS_MULTI_STATUS = 207;
    const STATUS_ALREADY_REPORTED = 208;
    const STATUS_IM_USED = 226;
    // Redirection 3xx
    const STATUS_MULTIPLE_CHOICES = 300;
    const STATUS_MOVED_PERMANENTLY = 301;
    const STATUS_FOUND = 302;
    const STATUS_SEE_OTHER = 303;
    const STATUS_NOT_MODIFIED = 304;
    const STATUS_USE_PROXY = 305;
    const STATUS_RESERVED = 306;
    const STATUS_TEMPORARY_REDIRECT = 307;
    const STATUS_PERMANENT_REDIRECT = 308;
    // Client Errors 4xx
    const STATUS_BAD_REQUEST = 400;
    const STATUS_UNAUTHORIZED = 401;
    const STATUS_PAYMENT_REQUIRED = 402;
    const STATUS_FORBIDDEN = 403;
    const STATUS_NOT_FOUND = 404;
    const STATUS_METHOD_NOT_ALLOWED = 405;
    const STATUS_NOT_ACCEPTABLE = 406;
    const STATUS_PROXY_AUTHENTICATION_REQUIRED = 407;
    const STATUS_REQUEST_TIMEOUT = 408;
    const STATUS_CONFLICT = 409;
    const STATUS_GONE = 410;
    const STATUS_LENGTH_REQUIRED = 411;
    const STATUS_PRECONDITION_FAILED = 412;
    const STATUS_PAYLOAD_TOO_LARGE = 413;
    const STATUS_URI_TOO_LONG = 414;
    const STATUS_UNSUPPORTED_MEDIA_TYPE = 415;
    const STATUS_RANGE_NOT_SATISFIABLE = 416;
    const STATUS_EXPECTATION_FAILED = 417;
    const STATUS_IM_A_TEAPOT = 418;
    const STATUS_MISDIRECTED_REQUEST = 421;
    const STATUS_UNPROCESSABLE_ENTITY = 422;
    const STATUS_LOCKED = 423;
    const STATUS_FAILED_DEPENDENCY = 424;
    const STATUS_TOO_EARLY = 425;
    const STATUS_UPGRADE_REQUIRED = 426;
    const STATUS_PRECONDITION_REQUIRED = 428;
    const STATUS_TOO_MANY_REQUESTS = 429;
    const STATUS_REQUEST_HEADER_FIELDS_TOO_LARGE = 431;
    const STATUS_UNAVAILABLE_FOR_LEGAL_REASONS = 451;
    // Server Errors 5xx
    const STATUS_INTERNAL_SERVER_ERROR = 500;
    const STATUS_NOT_IMPLEMENTED = 501;
    const STATUS_BAD_GATEWAY = 502;
    const STATUS_SERVICE_UNAVAILABLE = 503;
    const STATUS_GATEWAY_TIMEOUT = 504;
    const STATUS_VERSION_NOT_SUPPORTED = 505;
    const STATUS_VARIANT_ALSO_NEGOTIATES = 506;
    const STATUS_INSUFFICIENT_STORAGE = 507;
    const STATUS_LOOP_DETECTED = 508;
    const STATUS_NOT_EXTENDED = 510;
    const STATUS_NETWORK_AUTHENTICATION_REQUIRED = 511;
}
